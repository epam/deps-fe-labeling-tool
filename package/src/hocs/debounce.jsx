import React, { useEffect, useState, useCallback, useMemo, useRef } from 'react'
import lodashDebounce from 'lodash/debounce'
import PropTypes from 'prop-types'

const DEFAULT_WAIT = 250
const DEFAULT_GET_VALUE = (value) => value

const debounce = ({
  wait = DEFAULT_WAIT,
  getValue = DEFAULT_GET_VALUE
} = {}) => (ConnectedComponent) => {
  const Debounce = ({ onChange, value, ...rest }) => {
    const [currentValue, setCurrentValue] = useState(value)
    const mountedValue = useRef(value)

    useEffect(
      () => {
        if (value === mountedValue.current) {
          return
        }

        mountedValue.current = value
        setCurrentValue(value)
      },
      [value]
    )

    const debouncedOnChange = useMemo(
      () => lodashDebounce(
        (val) => onChange(val),
        wait
      ),
      [onChange]
    )

    const onChangeHandler = useCallback(
      (e) => {
        const value = getValue(e)
        setCurrentValue(value)
        debouncedOnChange(value)
      },
      [debouncedOnChange]
    )

    return (
      <ConnectedComponent
        {...rest}
        value={currentValue}
        onChange={onChangeHandler}
      />
    )
  }

  Debounce.propTypes = {
    value: PropTypes.any,
    onChange: PropTypes.func.isRequired
  }

  return Debounce
}

export {
  debounce
}
