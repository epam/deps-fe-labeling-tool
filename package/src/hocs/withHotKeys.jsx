import React, { useState, useCallback } from 'react'
import PropTypes from 'prop-types'
import { GlobalHotKeys, configure } from 'react-hotkeys'
import { connect } from 'react-redux'
import { registerHotKeyEvents } from '@/actions/hotkeys'
import { HotKeyEvent, customKeyCodes } from '@/constants/hotKeys'
import { hotKeysSelector } from '@/selectors/hotkeys'

configure({
  customKeyCodes
})

const generateKeyMap = (handlers) => {
  if (!handlers) {
    return
  }

  return Object.keys(handlers).reduce((acc, eventName) => {
    if (!eventName || !HotKeyEvent[eventName]) {
      return acc
    }

    return {
      ...acc,
      [eventName]: HotKeyEvent[eventName].shortcut
    }
  }, {})
}

const withHotKeys = (ConnectedComponent) => {
  const HotKeys = ({ registredHotKeyEvents, registerHotKeyEvents, ...props }) => {
    const [handlers, setHandlers] = useState()
    const [keyMap, setKeyMap] = useState()

    const registerHandlers = useCallback((handlers) => {
      const newEvents = Object.keys(handlers)
      if (newEvents.some((e) => !registredHotKeyEvents.includes(e))) {
        registerHotKeyEvents(newEvents)
      }

      const keyMap = generateKeyMap(handlers)
      setKeyMap(keyMap)
      setHandlers(handlers)
    }, [
      registredHotKeyEvents,
      registerHotKeyEvents
    ])

    const registerModifiers = useCallback((modifiers) => {
      registerHotKeyEvents(modifiers)
    }, [registerHotKeyEvents])

    return (
      <GlobalHotKeys
        keyMap={keyMap}
        handlers={handlers}
        allowChanges
      >
        <ConnectedComponent
          {...props}
          registerHandlers={registerHandlers}
          registerModifiers={registerModifiers}
        />
      </GlobalHotKeys>
    )
  }

  HotKeys.propTypes = {
    registerHotKeyEvents: PropTypes.func.isRequired,
    registredHotKeyEvents: PropTypes.arrayOf(PropTypes.string).isRequired
  }

  const mapStateToProps = (state) => ({
    registredHotKeyEvents: hotKeysSelector(state)
  })

  const mapDispatchToProps = {
    registerHotKeyEvents
  }

  return connect(mapStateToProps, mapDispatchToProps)(HotKeys)
}

export {
  withHotKeys
}
