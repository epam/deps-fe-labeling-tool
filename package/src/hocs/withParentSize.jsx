
import React from 'react'
import { withSize } from 'react-sizeme'
import styled from 'styled-components'

const FullSize = styled.div`
  width: 100%;
  height: 100%;
  overflow: hidden;
`

const withParentSize = (options) => (ConnectedComponent) => withSize(options)((props) => (
  <FullSize>
    <ConnectedComponent {...props} />
  </FullSize>
))

export {
  withParentSize
}
