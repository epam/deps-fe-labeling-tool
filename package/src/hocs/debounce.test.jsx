import React from 'react'
import { shallow } from 'enzyme'
import { debounce } from '@/hocs/debounce'

const MockComponent = (props) => <div {...props}>test</div>

const mockLodashDebounce = jest.fn((f) => f)
jest.mock('lodash/debounce', () => jest.fn((...arg) => mockLodashDebounce(...arg)))

describe('HOC: debounce', () => {
  let defaultProps, wrapper, DebouncedMockComponent

  const wait = 100
  const getValue = jest.fn((value) => value + 1)

  beforeEach(() => {
    defaultProps = {
      onChange: jest.fn(),
      value: 'test',
      testProps: 1
    }

    DebouncedMockComponent = debounce()(MockComponent)

    wrapper = shallow(<DebouncedMockComponent {...defaultProps} />)
  })

  it('should renders correctly', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should call props.onChange with correct arg', () => {
    wrapper.props().onChange('hi')
    expect(defaultProps.onChange).nthCalledWith(1, 'hi')
  })

  it('should correctly transmit data', () => {
    wrapper.props().onChange('hi')
    expect(wrapper.props().value).toEqual('hi')
  })

  it('should call getValue arg in case calling to passed component prop onChange', () => {
    DebouncedMockComponent = debounce({ wait, getValue })(MockComponent)

    wrapper = shallow(<DebouncedMockComponent {...defaultProps} />)

    wrapper.props().onChange('hi')
    expect(getValue).toBeCalledWith('hi')
  })

  it('should called lodashDebounce with correctly arg', () => {
    DebouncedMockComponent = debounce({ wait, getValue })(MockComponent)

    wrapper = shallow(<DebouncedMockComponent {...defaultProps} />)

    wrapper.props().onChange('hi')
    expect(mockLodashDebounce).nthCalledWith(2, expect.any(Function), wait)
  })
})
