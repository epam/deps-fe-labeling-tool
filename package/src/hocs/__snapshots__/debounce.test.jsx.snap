// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HOC: debounce should renders correctly 1`] = `
<MockComponent
  onChange={[Function]}
  testProps={1}
  value="test"
/>
`;
